lib/streamlit/errors.py

```diff
@@ -158,3 +161,284 @@ def __init__(self, module_name, *args):
+    def __init__(self, format: str):
+        super().__init__(
+            "Format string for st.number_input contains invalid characters: {format}",
+            format=format,
+        )
+
+
+# st.page_link
+class StreamlitMissingPageLabelError(LocalizableStreamlitException):
+    """Exception raised when a page_link is created without a label."""
+
+    def __init__(self):
+        super().__init__(
+            "The `label` param is required for external links used with `st.page_link` - please provide a `label`."
+        )
+
+
+class StreamlitPageNotFoundError(LocalizableStreamlitException):
+    """Exception raised the linked page can not be found."""
+
+    def __init__(self, page: str, main_script_directory: str, is_mpa_v2: bool):
+        directory = os.path.basename(main_script_directory)
+        message = None
+        if is_mpa_v2:
+            message = (
+                "Could not find page: `{page}`. You must provide a `StreamlitPage` "
+                "object or file path relative to the entrypoint file. Only pages "
+                "previously defined by [st.Page](http://st.page/) and passed to "
+                "`st.navigation` are allowed."
+            )
+        else:
+            message = (
+                "Could not find page: `{page}`. You must provide a file path "
+                "relative to the entrypoint file (from the directory `{directory}`). "
+                "Only the entrypoint file and files in the `pages/` directory are supported.",
+            )
+
+        super().__init__(
+            message,
+            page=page,
+            directory=directory,
+        )
+
+
+class StreamlitPageNotFoundMPAV2Error(LocalizableStreamlitException):
+    """Exception raised the linked page can not be found."""
+
+    def __init__(self, page: str):
+        super().__init__(
+            "Could not find page: `{page}`. You must provide a `StreamlitPage` "
+            "object or file path relative to the entrypoint file. Only pages "
+            "previously defined by [st.Page](http://st.page/) and passed to "
+            "`st.navigation` are allowed.",
+            page=page,
+        )
+
+
+# policies
+class StreamlitFragmentWidgetsNotAllowedOutsideError(LocalizableStreamlitException):
+    """Exception raised when the fragment attempts to write to an element outside of its container."""
+
+    def __init__(self):
+        super().__init__("Fragments cannot write widgets to outside containers.")
+
+
+class StreamlitInvalidFormCallbackError(LocalizableStreamlitException):
+    """Exception raised a `on_change` callback is set on any element in a form except for the `st.form_submit_button`."""
+
+    def __init__(self):
+        super().__init__(
+            "Within a form, callbacks can only be defined on st.form_submit_button. "
+            "Defining callbacks on other widgets inside a form is not allowed."
+        )
+
+
+class StreamlitWidgetValueAssignmentNotAllowedError(LocalizableStreamlitException):
```

Comment:
Seeing this on a second readthrough. Since we took "Widget" out of `StreamlitWidgetValueAboveMaxError` (and similarly for min), do we want to generalize this as well? e.g. `StreamlitValueAssignmentNotAllowedError` I can see a case for that since if we extend the use of keys a bit more we might end up with keys on non-widgets and want to raise the same error, "Hey you can't set this value."

Reply:
Sure!